Clean up and improve the code and public interface

llucax opened the pull request:
This PR addresses many small cleanups and improvements to the code and the public interface, including some minor breaking changes.

Cleanups:

- Use `deque` for typing
- Replace `__anext__()` with `anext()`
- Rename `recv` argument as `receiver`
- Add missing types to members in `__init__()`
- Anycast: Don't store the `_limit`
- Use the string representation for logging
- Broadcast: Move the `resend_latest` attribute to the end
- Remove `logging` from `Timer` examples
- Make `Broadcast` logger private

Improvements:

- Add `is_closed` property to channels
- Anycast: Clarify what happens when the buffer is full
- Anycast: Add `name` argument
- Add `name` property to bidirectional
- Broadcast: Use `hash()` instead of a UUID
- Event: Make the default name more readable
- Add descriptive `str` and `repr` implementations
- Bump `repo-config` to v0.7.5
- Bump `pylint` to 3.0.2
- Bump `mike` to 2.0.0
- Anycast: Log a warning when a sender is blocked

Breaking changes:

- Make some arguments keyword-only
- Make exception messages `str`
- Rename `maxsize` to `limit`
- Broadcast: Make `name` keyword-only
- Bidirectional: unify `client_id` and `service_id` (fixes #187).

llucax committed:
Make some arguments keyword-only

We do this when the arguments could be easily mistake or hard to know
what they are for in the call site, for example when the type is a
`str` or `int`, and for most optional arguments.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Make exception messages `str`

There is no reason to have them of `Any` type, anybody can explicitly
convert any object to `str` using `str(obj)` if they want to do so.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Replace `__anext__()` with `anext()`

`anext()` wasn't available in older Python versions.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Anycast: Don't store the `_limit`

The `_limit` is already accesible via the `_deque`'s `maxsize`.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Rename `maxsize` to `limit`

`maxsize` comes from `deque` and it is not very clear, and doesn't
follow the snake_case convention.

Also exposes the `limit` for `Anycast` as a (read-only) property.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Anycast: Clarify what happens when the buffer is full

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add `is_closed` property to channels

It might be useful to be able to test if a channel is closed or not.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Anycast: Add a `name` argument

This is for logging purposes, and to show in the string representation.
It is added to match other channels even when no logging is emitted in
this channel yet.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Broadcast: Make `name` optional and keyword-only

Since the `name` is only for debugging purposes, it shouldn't be
required. If no `name` is specified, a name will be created based on the
`id()` of the channel, so channels can be easily uniquely identified.

Also makes the `name` accessible via a read-only property.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Broadcast: Use `hash()` instead of a UUID

UUIDs are expensive to create and not really necessary in this case as
we are just keeping a local list of unique objects, so using the
object's hash is enough.

We also use a `id(self)`-based name by default if a `name` was not
provided when creating a `new_receiver()`.

A simple test using `timeit` shows a 2 orders of magnitude improvement
in Python 3.11:

```console
$ python -m timeit -s "import uuid" "uuid.uuid4()"
100000 loops, best of 5: 2.67 usec per loop
$ python -m timeit "hash(__name__)"
10000000 loops, best of 5: 19.6 nsec per loop
$ python -m timeit "id(__name__)"
10000000 loops, best of 5: 20.1 nsec per loop
```

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add missing types to members in `__init__()`

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Event: Make the default name more readable

Use underscore separators to format the `id(self)`. Also only use the
default if `name` is `None`. Before an empty string would also be
changed to the default, but if an user passed an empty string, it is
better to leave it untouched.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Use the string representation for logging

Now that we have nice string representations, it is no longer needed to
build one every time we log something.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Bidirectional: unify `client_id` and `service_id`

This is only used to create the underlying `Broadcast` channel's names,
so instead of using 2 separate strings, just use a plain `name` as with
other channels.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Rename `recv` argument as `receiver`

For the rest of the API we use full names instead of abbreviated names,
so better to do the same here for consistency and clarity.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add `name` property to bidirectional

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add descriptive `str` and `repr` implementations

When debugging and logging objects it is very useful to get a
descriptive and clear string representation.

The new representation uses the class name and the user defined name
(if any) for `str` and a `repr` that tries to show how the class was
created but also important internal state.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Use `deque` for typing

There is no need to import `typing.Deque` in newer Python versions.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Broadcast: Move the `resend_latest` attribute to the end

This is so it is closer to the other public attributes (properties) of
the class.

Also remove the mention to the default, as the default belongs to the
`__init__` argument docs.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Bump `repo-config` to v0.7.5

- Adds support for `pylint` 3.
- Adds support for `mike` 2.
- Fixes a bug in mike version sorting.
- Make documentation publishing job will not fail if the version for
  `mike` can't be determined, it will just emit a warning and skip the
  publishing of the documentation website.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Bump `pylint` to 3.0.2

A few new issues are detected, which are also fixed in this commit.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Bump `mike` to 2.0.0

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

shsms commented on src/frequenz/channels/_broadcast.py:
```diff
@@ -161,15 +159,15 @@ def new_receiver(
         Args:
             name: A name to identify the receiver in the logs.
-            maxsize: Size of the receiver's buffer.
+            limit: Size of the receiver's buffer in number of messages.
```
Maybe this is better?

```suggestion
            limit: Number of messages the receiver can hold in its buffer.

llucax replied:
Updated with the suggestion.

shsms commented on src/frequenz/channels/_broadcast.py:
```diff
@@ -71,12 +71,13 @@ async def recv(id: int, receiver: channel.Receiver) -> None:
     """
 
-    def __init__(self, name: str, *, resend_latest: bool = False) -> None:
+    def __init__(self, *, name: str | None, resend_latest: bool = False) -> None:
```
I'm not so sure about making the name parameter of `Broadcast` channels optional, because people would just create channels without thinking of names, and we'll have logs complaining about random channels filling up, and it becomes super hard to find out where the bottle neck is.

christianparpart replied:
Why do you prefer `A | None` over `Optional[A]`? I am asking purely out of curiosity. :)

llucax replied:
Not really strong reasons, but no need for an extra `import` and `Optional` was just a library solution until `|` got properly implemented in Python 3.10.

llucax replied:
> I'm not so sure about making the name parameter of `Broadcast` channels optional, because people would just create channels without thinking of names, and we'll have logs complaining about random channels filling up, and it becomes super hard to find out where the bottle neck is.

Mmm, I see. For me it was weird to require an argument that it is only used for logging, but I can see how otherwise people might need to learn the hard way to use custom names.

Also my feeling is, this is only a special case now in `Broadcast` because is what we use the most, but following that argument, `Anycast` should probably also log when a sender is being blocked because the buffer is full, as it is also a sign of receivers not being able to keep up.

Also, this might apply to other classes too, like `Merge`or whatever, where you could potentially have issues with one instance and that get logged and then you don't know which one is giving issues (although I know this is more a theoretical issue). But since this is an interface issue, and it is hard to know what future requirements would be, if names are really important we might think about making them required even if they are mostly unused for now (except for the `str`/`repr` implementation).

I know the above might be a bit extreme, but at least for channel classes, I think we should have a consistent approach and have all names optional or required.

shsms replied:
Yes, I agree on consistency, and we need it to be mandatory only for the channels, because everything else can derive the names for logging from the channel.

llucax replied:
OK, I will make it mandatory for all channels then.

llucax replied:
Done

shsms commented on src/frequenz/channels/_broadcast.py:
```diff
@@ -127,6 +127,15 @@ def name(self) -> str:
         return self._name
 
+    @property
+    def latest(self) -> T | None:
```
Not sure it is a good idea to add this directly on the channel, unless you have a clear usecase in mind already?

llucax replied:
No, not in particular, it just entered in the bag of private stuff that should be safe to expose as read-only. But thinking about the case we were talking today about only needing the last value, I was sort of wondering why even a receiver is needed at all if we can access the `latest` value from the channel directly.

I guess your answer will be something like "because the channel is sort of a private interface and should be used through senders and receivers", not sure what's the right balance between pragmatism and purism :D

shsms replied:
> I was sort of wondering why even a receiver is needed at all if we can access the latest value from the channel directly.

yes, that's exactly what a `Peekable` does.  And the `_latest` variable was added to support the `Peekable` implementation, so I'd rather remove that as well.

And not all channels are accessible.  We have the `new_receiver` interface everywhere in the sdk's data pipeline, and even in the microgrid client.  Those classes own the channels and give out only receivers.  So it is also easier to have the same interface everywhere.

shsms replied:
And it is one less hurdle once we start trying to go to multi-process world.

llucax replied:
OK, I will remove it.

llucax replied:
Done

shsms reviewed (commented):
Just a few comments.

Also looks like some commented lines got into the commit message here, I wonder how: https://github.com/frequenz-floss/frequenz-channels-python/pull/231/commits/25a9099f6ebd79046a215281bf6c3c70a5d6311d

llucax commented:
> Also looks like some commented lines got into the commit message here, I wonder how: [25a9099](https://github.com/frequenz-floss/frequenz-channels-python/commit/25a9099f6ebd79046a215281bf6c3c70a5d6311d)

Oh, good catch, thanks, it sometimes happen when using the scissors style for comments, I use it to avoid messing with markdown in Git messages.

llucax commented:
> Oh, good catch, thanks, it sometimes happen when using the scissors style for comments, I use it to avoid messing with markdown in Git messages.

Fixed in the update too.

llucax committed:
Make `Broadcast` logger private

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Anycast: Log a warning when a sender is blocked

This could indicate receivers can't keep up, so it is better to have
some warning about this.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove `logging` from `Timer` examples

The `logging` module wasn't properly used (no explicit creation of a
logger) and there it was mixed with prints. Using `logging` only makes
the example more complicated.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Updated to remove the `Broadcast.latest` property and to make channel names required.

Also added these 3 extra commits cleaning up logging a bit and adding a new log when `Anycast` senders are blocked as previously suggested:

* Remove `logging` from `Timer` examples
* `Anycast`: Log a warning when a sender is blocked
* Make `Broadcast` logger private
